### PR TITLE
Fix possible alignement issues on 32 bits CPUs

### PR DESCRIPTION
**Pull request by VictorLamoine**
https://github.com/PointCloudLibrary/pcl/pull/1387#issuecomment-149149884

`apps_cloud_composer` still compiles fine.

### CODE REVIEW

**Comment by taketwo** [comment]
Hi Victor, thanks for taking care of this. I think that we don't need to get rid of `make_shared` everywhere. In fact, it's slightly [more efficient](http://stackoverflow.com/a/20895705/1525865) and the syntax is definitely more succinct. So I'd vote for keeping `make_shared` where aligned objects are not involved.

**Comment by SergioRAgostinho** [comment]
Can we integrate this in 1.8.1? I'm asking because of that WIP tag :)

File: apps/cloud_composer/include/pcl/apps/cloud_composer/impl/merge_selection.hpp [hunk 1]
```diff
@@ -79,18 +79,18 @@ pcl::cloud_composer::MergeSelection::performTemplatedAction (QList <const CloudC
...
       
       qDebug () << "Original minus indices is "<<original_minus_indices->width;
       
       //Eigen::Vector4f source_origin = input_cloud_item->data (ItemDataRole::ORIGIN).value<Eigen::Vector4f> ();
       //Eigen::Quaternionf source_orientation =  input_cloud_item->data (ItemDataRole::ORIENTATION).value<Eigen::Quaternionf> ();
-      //pcl::PCLPointCloud2::Ptr cloud_blob = boost::make_shared <pcl::PCLPointCloud2> ();;
+      //pcl::PCLPointCloud2::Ptr cloud_blob = boost::shared_ptr <pcl::PCLPointCloud2> (new pcl::PCLPointCloud2);
       //toPCLPointCloud2 (*original_minus_indices, *cloud_blob);
       //CloudItem* new_cloud_item = new CloudItem (input_cloud_item->text ()
                                                   //, cloud_blob
```
Comment (SergioRAgostinho) [inline apps/cloud_composer/include/pcl/apps/cloud_composer/impl/merge_selection.hpp]: Shouldn't we erase this comments all together?

File: apps/cloud_composer/include/pcl/apps/cloud_composer/tools/impl/organized_segmentation.hpp [hunk 2]
```diff
@@ -131,7 +132,7 @@
      {
        if (euclidean_label_indices[i].indices.size () >= min_cluster_size)
        {
-         typename PointCloud<PointT>::Ptr cluster = boost::make_shared <PointCloud<PointT> >();
+         typename PointCloud<PointT>::Ptr cluster = boost::shared_ptr <PointCloud<PointT> >(new PointCloud<PointT>);
```
Comment (SergioRAgostinho) [inline apps/cloud_composer/include/pcl/apps/cloud_composer/tools/impl/organized_segmentation.hpp]: Missing space before the parenthesis.

File: apps/cloud_composer/src/items/cloud_item.cpp [hunk 3]
```diff
@@ -143,31 +143,31 @@ pcl::cloud_composer::CloudItem::setTemplateCloudFromBlob ()
     {
       case (PointTypeFlags::XYZ):
       {
-        pcl::PointCloud <PointXYZ>::Ptr cloud_ptr =  boost::make_shared<pcl::PointCloud <PointXYZ> >();
+        pcl::PointCloud <PointXYZ>::Ptr cloud_ptr = boost::shared_ptr<pcl::PointCloud <PointXYZ> > (new pcl::PointCloud <PointXYZ> );
         pcl::fromPCLPointCloud2 (*cloud_blob_ptr_, *cloud_ptr);
         cloud_pointer_variant = QVariant::fromValue (cloud_ptr);
         //Initialize the search kd-tree for this cloud
-        pcl::search::KdTree<PointXYZ>::Ptr kd_search = boost::make_shared<search::KdTree<PointXYZ> >();
+        pcl::search::KdTree<PointXYZ>::Ptr kd_search = boost::shared_ptr<search::KdTree<PointXYZ> >(new search::KdTree<PointXYZ>);
```
Comment (SergioRAgostinho) [inline apps/cloud_composer/src/items/cloud_item.cpp]: Missing space before the parenthesis. Change also line 157, 160, 167 and 170.

File: apps/cloud_composer/src/project_model.cpp [hunk 4]
```diff
@@ -106,7 +106,7 @@ pcl::cloud_composer::ProjectModel::setPointSelection (boost::shared_ptr<Selectio
   // Find all indices in the selected points which are present in the clouds
   foreach (CloudItem* cloud_item, project_clouds)
   {
-    pcl::PointIndices::Ptr found_indices = boost::make_shared<pcl::PointIndices>();
+    pcl::PointIndices::Ptr found_indices = boost::shared_ptr<pcl::PointIndices>(new pcl::PointIndices);
```
Comment (SergioRAgostinho) [inline apps/cloud_composer/src/project_model.cpp]: Missing space before the parenthesis. Also in line 277.

**Comment by SergioRAgostinho** [comment]
In any case double check the spaces before the parenthesis and I would say to erase those comment blocks unless someone opposes.

**Comment by VictorLamoine** [comment]
> So I'd vote for keeping make_shared where aligned objects are not involved.

How do I know why objects are aligned/not aligned?

**Comment by taketwo** [comment]
Aligned are objects that have Eigen members inside. Looking at the changes in this PR, I think only  `pcl::PointCloud<T>`s are aligned. Everything else is not (note: `PCLPointCloud2` are _not_ aligned).

**Comment by VictorLamoine** [comment]
Ready for review

**Comment by taketwo** [comment]
:+1:

**Comment by SergioRAgostinho** [comment]
Is there a way for us to prevent people from using make_shared with PointCloudT types (overriding?) or will we just need to pay attention to all subsequent PRs?

**Comment by taketwo** [comment]
Here is a brief explanation of what is wrong with `make_shared` and aligned objects (from [mailing list](http://www.pcl-users.org/Can-t-avoid-Eigen-alignment-problems-td2531193.html)):

> Don't use make_shared() with objects that need to be aligned.  [1] 
> The longer version is that make_shared() allocates the memory itself, 
> and then calls "placement new" [2] to initialize the object.  By 
> allocating the memory itself, make_shared() circumvents the "operator 
> new" that EIGEN_MAKE_ALIGNED_OPERATOR_NEW creates for you.  This is 
> similar to how std::vector works, and why you need to pass it an 
> allocator when using it with types that must be aligned. 
> [1] http://lists.boost.org/boost-users/2010/03/57713.php
> [2] http://en.wikipedia.org/wiki/Placement_syntax

Now the solution might be to disable or make private the placement new operator, so that `make_shared` can not call it. Is this possible? No idea.

**Pull request merged**